clients/src/main/java/org/apache/kafka/common/serialization/ListSerializer.java

```diff
@@ -0,0 +1,98 @@
+import org.apache.kafka.common.utils.Utils;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+public class ListSerializer<Inner> implements Serializer<List<Inner>> {
+
+    private Serializer<Inner> inner;
+    private boolean isFixedLength;
+
+    static private List<Class<? extends Serializer<?>>> fixedLengthSerializers = Arrays.asList(
+        ShortSerializer.class,
+        IntegerSerializer.class,
+        FloatSerializer.class,
+        LongSerializer.class,
+        DoubleSerializer.class,
+        UUIDSerializer.class);
+
+    public ListSerializer() {}
+
+    public ListSerializer(Serializer<Inner> serializer) {
+        this.inner = serializer;
+        this.isFixedLength = fixedLengthSerializers.contains(serializer.getClass());
+    }
+
+    @SuppressWarnings(value = "unchecked")
+    @Override
+    public void configure(Map<String, ?> configs, boolean isKey) {
+        if (inner == null) {
+            final String innerSerializerPropertyName = isKey ? CommonClientConfigs.DEFAULT_LIST_KEY_SERIALIZER_INNER_CLASS : CommonClientConfigs.DEFAULT_LIST_VALUE_SERIALIZER_INNER_CLASS;
+            final Object innerSerializerClassOrName = configs.get(innerSerializerPropertyName);
+            try {
+                if (innerSerializerClassOrName instanceof String) {
+                    inner = Utils.newInstance((String) innerSerializerClassOrName, Serializer.class);
+                } else if (innerSerializerClassOrName instanceof Class) {
+                    inner = Utils.newInstance((Class<Serializer<Inner>>) innerSerializerClassOrName);
+                } else {
+                    throw new ClassNotFoundException();
+                }
+                inner.configure(configs, isKey);
+            } catch (final ClassNotFoundException e) {
+                throw new ConfigException(innerSerializerPropertyName, innerSerializerClassOrName, "Serializer class " + innerSerializerClassOrName + " could not be found.");
+            }
+        }
+    }
+
+    @Override
+    public byte[] serialize(String topic, List<Inner> data) {
+        if (data == null) {
+            return null;
+        }
+        final int size = data.size();
+        try (final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+             final DataOutputStream out = new DataOutputStream(baos)) {
+            out.writeInt(size);
+            for (Inner entry : data) {
+                final byte[] bytes = inner.serialize(topic, entry);
+                if (!isFixedLength) {
+                    out.writeInt(bytes.length);
+                }
+                out.write(bytes);
+            }
+            return baos.toByteArray();
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to serialize List", e);
+        }
+    }
+
+    @Override
+    public void close() {
+        inner.close();
```

Comment:
We should add a `null` check to allow closing a serializer that was not properly setup